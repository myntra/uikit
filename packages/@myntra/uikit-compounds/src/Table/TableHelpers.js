import React from 'react'

function makeRender(accessor) {
  if (typeof accessor === 'function') {
    return ({ data }) => accessor(data)
  }

  return ({ data }) => data[accessor]
}

export function processColumn(depth, column) {
  if (!column.key) throw new Error('Table.Column should be keyed.')

  const { children: ref, label, accessor, ...props } = column.props
  const isRenderFunction = ref && typeof ref === 'function'
  const render = isRenderFunction
    ? ref
    : React.Children.count(ref) === 0
      ? makeRender(accessor || column.key)
      : undefined
  const children = render
    ? []
    : React.Children.count(ref) > 1
      ? React.Children.map(ref, item => processColumn(depth + 1, item))
      : [processColumn(depth + 1, ref)]

  return {
    key: column.key,
    depth: depth,
    render,
    children,
    props: {
      ...props,
      key: column.key,
      children: label,
      colSpan: children.reduce((acc, child) => acc + child.props.colSpan, 0) || 1
    }
  }
}
export function processColumns(columns) {
  return React.Children.map(columns, column => processColumn(1, column))
}

function renderTableHeader(columns, order) {
  const items = Object.values(columns)
  const depth = items.reduce((acc, column) => Math.max(acc, column.depth), 1)
  const rows = Array.apply(null, { length: depth }).map(() => [])
  const orderedColumns = []
  const generateHeaders = (i, column) => {
    rows[i].push(column.props)

    if (column.children.length) {
      column.children.forEach(child => generateHeaders(i + 1, child))
    } else {
      column.props.rowSpan = depth - i
      orderedColumns.push(column)
    }
  }

  order =
    Array.isArray(order) && order.length
      ? order.filter(key => key && columns[key] && columns[key].depth === 1)
      : items.filter(column => column.depth === 1).map(column => column.key)

  order.forEach(key => generateHeaders(0, columns[key]))

  return { order: orderedColumns, rows: rows.map((columns, id) => ({ id, columns })), colgroup: rows[0] }
}

export function prepareHead(ref, order) {
  const columns = {}
  const addToMap = column => {
    if (column.key in columns) throw new Error(`Duplicate key '${column.key}' found in table`)
    columns[column.key] = column
    column.children.forEach(addToMap)
  }

  // Process.
  processColumns(ref).forEach(addToMap)

  return { columns, ...renderTableHeader(columns, order) }
}

function prepareCellProps({ rowSpan, colSpan, transformProps, ...props }) {
  if (typeof transformProps === 'function') {
    return transformProps(props)
  }

  return props
}

export function prepareBody(head, data, id) {
  return {
    rows: data.map((row, index) => ({
      id: id(row, index),
      columns: head.order.map(column => ({
        ...prepareCellProps(column.props),
        key: column.key,
        children: column.render({ data: row, column, meta: column.props })
      }))
    }))
  }
}
