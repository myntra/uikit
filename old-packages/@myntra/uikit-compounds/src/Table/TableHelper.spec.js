import React from 'react'
import { mount } from 'enzyme'
import { prepareBody, prepareHead } from './TableHelpers'
import Column from './TableColumn'

function prepareForTest(done, fn) {
  return ({ children, order, data }) => {
    const meta = prepareHead(children, order)

    expect(meta).toMatchSnapshot()

    if (data) {
      const body = prepareBody(meta, data, (row, index) => index)

      expect(body).toMatchSnapshot()

      fn && fn(body)
    } else {
      fn && fn(meta)
    }

    done()

    return null
  }
}

it('should collect header info', done => {
  const Component = prepareForTest(done, meta => {
    expect(meta.rows).toHaveLength(1)
    expect(meta.rows[0].columns).toHaveLength(3)
    expect(meta.rows[0].columns.map(column => column.children)).toEqual(['ID', 'Name', 'Content'])
  })

  mount(
    <Component>
      <Column key="id" label="ID" />
      <Column key="name" label="Name" accessor={({ firstName, lastName }) => `${firstName} ${lastName}`} />
      <Column key="content" label="Content">
        {({ data }) => <span>data.content</span>}
      </Column>
    </Component>
  )
})

it('should allow header grouping', done => {
  const Component = prepareForTest(done, meta => {
    expect(meta.rows).toHaveLength(2)
    expect(meta.rows[0].columns).toHaveLength(3)
    expect(meta.rows[0].columns.map(column => column.children)).toEqual(['ID', 'Name', 'Content'])
    expect(meta.rows[0].columns[0].colSpan).toBe(1)
    expect(meta.rows[0].columns[1].colSpan).toBe(2)
    expect(meta.rows[1].columns).toHaveLength(2)
    expect(meta.rows[1].columns.map(column => column.children)).toEqual(['First Name', 'Last Name'])
  })

  mount(
    <Component>
      <Column key="id" label="ID" />
      <Column key="name" label="Name">
        <Column key="firstName" label="First Name" />
        <Column key="lastName" label="Last Name" />
      </Column>
      <Column key="content" label="Content" />
    </Component>
  )
})

it('should throw error on missing key', done => {
  const Component = ({ children, order, data }) => {
    expect(() => prepareHead(children, order)).toThrow('Table.Column should')

    done()

    return null
  }

  mount(
    <Component>
      <Column key="id" label="ID" />
      <Column key="name" label="Name">
        <Column label="First Name" />
        <Column key="lastName" label="Last Name" />
      </Column>
      <Column key="content" label="Content" />
    </Component>
  )
})

it('should throw error on unknown key in order', done => {
  const Component = ({ children, order, data }) => {
    expect(() => prepareHead(children, order)).toThrow('Duplicate key')

    done()

    return null
  }

  mount(
    <Component order={['foo']}>
      <Column key="id" label="ID" />
      <Column key="name" label="Name">
        <Column key="name" label="Last Name" />
      </Column>
      <Column key="content" label="Content" />
    </Component>
  )
})

it('should allow column to have one child', done => {
  const Component = prepareForTest(done, meta => {
    expect(meta.rows).toHaveLength(2)
    expect(meta.rows[0].columns).toHaveLength(3)
    expect(meta.rows[0].columns.map(column => column.children)).toEqual(['ID', 'Name', 'Content'])
    expect(meta.rows[0].columns[0].colSpan).toBe(1)
    expect(meta.rows[0].columns[1].colSpan).toBe(1)
    expect(meta.rows[1].columns).toHaveLength(1)
    expect(meta.rows[1].columns.map(column => column.children)).toEqual(['First Name'])
  })

  mount(
    <Component>
      <Column key="id" label="ID" />
      <Column key="name" label="Name">
        <Column key="firstName" label="First Name" />
      </Column>
      <Column key="content" label="Content" />
    </Component>
  )
})

it('should reorder columns', done => {
  const Component = prepareForTest(done, meta => {
    expect(meta.rows).toHaveLength(1)
    expect(meta.rows[0].columns).toHaveLength(2)
    expect(meta.rows[0].columns.map(column => column.children)).toEqual(['Name', 'ID'])
  })

  mount(
    <Component order={['name', 'id']}>
      <Column key="id" label="ID" />
      <Column key="name" label="Name" accessor={({ firstName, lastName }) => `${firstName} ${lastName}`} />
      <Column key="content" label="Content">
        {({ data }) => <span>data.content</span>}
      </Column>
    </Component>
  )
})

it('should prepare table rows', done => {
  const Component = prepareForTest(done, body => {
    expect(body.rows).toHaveLength(1)
    expect(body.rows[0].columns).toHaveLength(3)
    expect(body.rows[0].columns.map(column => column.children)).toEqual([1, 'John Doe', <span>foo</span>]) // eslint-disable-line react/jsx-key
  })

  const fn = jest.fn().mockImplementation(props => props)

  mount(
    <Component data={[{ id: 1, firstName: 'John', lastName: 'Doe', content: 'foo' }]}>
      <Column key="id" label="ID" />
      <Column key="name" label="Name" accessor={({ firstName, lastName }) => `${firstName} ${lastName}`} />
      <Column key="content" label="Content" transformProps={fn}>
        {({ data }) => <span>{data.content}</span>}
      </Column>
    </Component>
  )

  expect(fn).toHaveBeenCalled()
})
