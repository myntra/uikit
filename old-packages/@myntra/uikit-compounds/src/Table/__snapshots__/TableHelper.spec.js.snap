// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow column to have one child 1`] = `
Object {
  "colgroup": Array [
    Object {
      "children": "ID",
      "colSpan": 1,
      "key": "id",
      "rowSpan": 2,
    },
    Object {
      "children": "Name",
      "colSpan": 1,
      "key": "name",
    },
    Object {
      "children": "Content",
      "colSpan": 1,
      "key": "content",
      "rowSpan": 2,
    },
  ],
  "columns": Object {
    "content": Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
        "rowSpan": 2,
      },
      "render": [Function],
    },
    "firstName": Object {
      "children": Array [],
      "depth": 2,
      "key": "firstName",
      "props": Object {
        "children": "First Name",
        "colSpan": 1,
        "key": "firstName",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    "id": Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 2,
      },
      "render": [Function],
    },
    "name": Object {
      "children": Array [
        Object {
          "children": Array [],
          "depth": 2,
          "key": "firstName",
          "props": Object {
            "children": "First Name",
            "colSpan": 1,
            "key": "firstName",
            "rowSpan": 1,
          },
          "render": [Function],
        },
      ],
      "depth": 1,
      "key": "name",
      "props": Object {
        "children": "Name",
        "colSpan": 1,
        "key": "name",
      },
      "render": undefined,
    },
  },
  "order": Array [
    Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 2,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 2,
      "key": "firstName",
      "props": Object {
        "children": "First Name",
        "colSpan": 1,
        "key": "firstName",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
        "rowSpan": 2,
      },
      "render": [Function],
    },
  ],
  "rows": Array [
    Object {
      "columns": Array [
        Object {
          "children": "ID",
          "colSpan": 1,
          "key": "id",
          "rowSpan": 2,
        },
        Object {
          "children": "Name",
          "colSpan": 1,
          "key": "name",
        },
        Object {
          "children": "Content",
          "colSpan": 1,
          "key": "content",
          "rowSpan": 2,
        },
      ],
      "id": 0,
    },
    Object {
      "columns": Array [
        Object {
          "children": "First Name",
          "colSpan": 1,
          "key": "firstName",
          "rowSpan": 1,
        },
      ],
      "id": 1,
    },
  ],
}
`;

exports[`should allow header grouping 1`] = `
Object {
  "colgroup": Array [
    Object {
      "children": "ID",
      "colSpan": 1,
      "key": "id",
      "rowSpan": 2,
    },
    Object {
      "children": "Name",
      "colSpan": 2,
      "key": "name",
    },
    Object {
      "children": "Content",
      "colSpan": 1,
      "key": "content",
      "rowSpan": 2,
    },
  ],
  "columns": Object {
    "content": Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
        "rowSpan": 2,
      },
      "render": [Function],
    },
    "firstName": Object {
      "children": Array [],
      "depth": 2,
      "key": "firstName",
      "props": Object {
        "children": "First Name",
        "colSpan": 1,
        "key": "firstName",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    "id": Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 2,
      },
      "render": [Function],
    },
    "lastName": Object {
      "children": Array [],
      "depth": 2,
      "key": "lastName",
      "props": Object {
        "children": "Last Name",
        "colSpan": 1,
        "key": "lastName",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    "name": Object {
      "children": Array [
        Object {
          "children": Array [],
          "depth": 2,
          "key": "firstName",
          "props": Object {
            "children": "First Name",
            "colSpan": 1,
            "key": "firstName",
            "rowSpan": 1,
          },
          "render": [Function],
        },
        Object {
          "children": Array [],
          "depth": 2,
          "key": "lastName",
          "props": Object {
            "children": "Last Name",
            "colSpan": 1,
            "key": "lastName",
            "rowSpan": 1,
          },
          "render": [Function],
        },
      ],
      "depth": 1,
      "key": "name",
      "props": Object {
        "children": "Name",
        "colSpan": 2,
        "key": "name",
      },
      "render": undefined,
    },
  },
  "order": Array [
    Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 2,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 2,
      "key": "firstName",
      "props": Object {
        "children": "First Name",
        "colSpan": 1,
        "key": "firstName",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 2,
      "key": "lastName",
      "props": Object {
        "children": "Last Name",
        "colSpan": 1,
        "key": "lastName",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
        "rowSpan": 2,
      },
      "render": [Function],
    },
  ],
  "rows": Array [
    Object {
      "columns": Array [
        Object {
          "children": "ID",
          "colSpan": 1,
          "key": "id",
          "rowSpan": 2,
        },
        Object {
          "children": "Name",
          "colSpan": 2,
          "key": "name",
        },
        Object {
          "children": "Content",
          "colSpan": 1,
          "key": "content",
          "rowSpan": 2,
        },
      ],
      "id": 0,
    },
    Object {
      "columns": Array [
        Object {
          "children": "First Name",
          "colSpan": 1,
          "key": "firstName",
          "rowSpan": 1,
        },
        Object {
          "children": "Last Name",
          "colSpan": 1,
          "key": "lastName",
          "rowSpan": 1,
        },
      ],
      "id": 1,
    },
  ],
}
`;

exports[`should collect header info 1`] = `
Object {
  "colgroup": Array [
    Object {
      "children": "ID",
      "colSpan": 1,
      "key": "id",
      "rowSpan": 1,
    },
    Object {
      "children": "Name",
      "colSpan": 1,
      "key": "name",
      "rowSpan": 1,
    },
    Object {
      "children": "Content",
      "colSpan": 1,
      "key": "content",
      "rowSpan": 1,
    },
  ],
  "columns": Object {
    "content": Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    "id": Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    "name": Object {
      "children": Array [],
      "depth": 1,
      "key": "name",
      "props": Object {
        "children": "Name",
        "colSpan": 1,
        "key": "name",
        "rowSpan": 1,
      },
      "render": [Function],
    },
  },
  "order": Array [
    Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 1,
      "key": "name",
      "props": Object {
        "children": "Name",
        "colSpan": 1,
        "key": "name",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
        "rowSpan": 1,
      },
      "render": [Function],
    },
  ],
  "rows": Array [
    Object {
      "columns": Array [
        Object {
          "children": "ID",
          "colSpan": 1,
          "key": "id",
          "rowSpan": 1,
        },
        Object {
          "children": "Name",
          "colSpan": 1,
          "key": "name",
          "rowSpan": 1,
        },
        Object {
          "children": "Content",
          "colSpan": 1,
          "key": "content",
          "rowSpan": 1,
        },
      ],
      "id": 0,
    },
  ],
}
`;

exports[`should prepare table rows 1`] = `
Object {
  "colgroup": Array [
    Object {
      "children": "ID",
      "colSpan": 1,
      "key": "id",
      "rowSpan": 1,
    },
    Object {
      "children": "Name",
      "colSpan": 1,
      "key": "name",
      "rowSpan": 1,
    },
    Object {
      "children": "Content",
      "colSpan": 1,
      "key": "content",
      "rowSpan": 1,
      "transformProps": [MockFunction],
    },
  ],
  "columns": Object {
    "content": Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
        "rowSpan": 1,
        "transformProps": [MockFunction],
      },
      "render": [Function],
    },
    "id": Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    "name": Object {
      "children": Array [],
      "depth": 1,
      "key": "name",
      "props": Object {
        "children": "Name",
        "colSpan": 1,
        "key": "name",
        "rowSpan": 1,
      },
      "render": [Function],
    },
  },
  "order": Array [
    Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 1,
      "key": "name",
      "props": Object {
        "children": "Name",
        "colSpan": 1,
        "key": "name",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
        "rowSpan": 1,
        "transformProps": [MockFunction],
      },
      "render": [Function],
    },
  ],
  "rows": Array [
    Object {
      "columns": Array [
        Object {
          "children": "ID",
          "colSpan": 1,
          "key": "id",
          "rowSpan": 1,
        },
        Object {
          "children": "Name",
          "colSpan": 1,
          "key": "name",
          "rowSpan": 1,
        },
        Object {
          "children": "Content",
          "colSpan": 1,
          "key": "content",
          "rowSpan": 1,
          "transformProps": [MockFunction],
        },
      ],
      "id": 0,
    },
  ],
}
`;

exports[`should prepare table rows 2`] = `
Object {
  "rows": Array [
    Object {
      "columns": Array [
        Object {
          "children": 1,
          "key": "id",
        },
        Object {
          "children": "John Doe",
          "key": "name",
        },
        Object {
          "children": <span>
            foo
          </span>,
          "key": "content",
        },
      ],
      "data": Object {
        "content": "foo",
        "firstName": "John",
        "id": 1,
        "lastName": "Doe",
      },
      "id": 0,
      "index": 0,
    },
  ],
}
`;

exports[`should reorder columns 1`] = `
Object {
  "colgroup": Array [
    Object {
      "children": "Name",
      "colSpan": 1,
      "key": "name",
      "rowSpan": 1,
    },
    Object {
      "children": "ID",
      "colSpan": 1,
      "key": "id",
      "rowSpan": 1,
    },
  ],
  "columns": Object {
    "content": Object {
      "children": Array [],
      "depth": 1,
      "key": "content",
      "props": Object {
        "children": "Content",
        "colSpan": 1,
        "key": "content",
      },
      "render": [Function],
    },
    "id": Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    "name": Object {
      "children": Array [],
      "depth": 1,
      "key": "name",
      "props": Object {
        "children": "Name",
        "colSpan": 1,
        "key": "name",
        "rowSpan": 1,
      },
      "render": [Function],
    },
  },
  "order": Array [
    Object {
      "children": Array [],
      "depth": 1,
      "key": "name",
      "props": Object {
        "children": "Name",
        "colSpan": 1,
        "key": "name",
        "rowSpan": 1,
      },
      "render": [Function],
    },
    Object {
      "children": Array [],
      "depth": 1,
      "key": "id",
      "props": Object {
        "children": "ID",
        "colSpan": 1,
        "key": "id",
        "rowSpan": 1,
      },
      "render": [Function],
    },
  ],
  "rows": Array [
    Object {
      "columns": Array [
        Object {
          "children": "Name",
          "colSpan": 1,
          "key": "name",
          "rowSpan": 1,
        },
        Object {
          "children": "ID",
          "colSpan": 1,
          "key": "id",
          "rowSpan": 1,
        },
      ],
      "id": 0,
    },
  ],
}
`;
